### ISSUE TITLE
storage: non-member replicas can be left around after a previously dead node restarts

### ISSUE THREAD

**Issue by a-robinson** [cockroachdb/cockroach#24067]
In https://github.com/cockroachdb/cockroach/issues/23994, we observed a restarted dead node completely failing to GC replicas that had been removed from their ranges because all of the other replicas in the ranges had been moved from where they were when the node was previously running.

https://github.com/cockroachdb/cockroach/pull/23995 fixed this for some of the replicas and got rid of the replicate queue backlog, but still leaves a case where such a restarted node won't GC some of its replicas. If the node that was epoch-based leaseholder for a given range when the node went down is still in the same epoch, the dead node will assume that the lease is still valid and not try to claim it for itself (which is how it would otherwise learn that everything has moved).

This isn't a huge deal, but in a very stable cluster where the other nodes aren't restarting or having their liveness epoch incremented, old replicas could be left around for quite a while.

**Comment by bdarnell** [comment]
> This isn't a huge deal, but in a very stable cluster where the other nodes aren't restarting or having their liveness epoch incremented, old replicas could be left around for quite a while.

These replicas can be reported as "underreplicated" and this can complicate efforts to cleanly automate node decommissioning and replacement.

I think the `if bq.needsLease` block in queue.go should have a no-op `withRaftGroup` call to wake up the raft group if it's not already. This will (eventually) trigger a replica GC queue if the replica should indeed be removed. If not, it should be cheap (since there would be heartbeats that would wake up the raft group soon anyway if it is alive)

**Comment by tschottdorf** [comment]
Related to #26952 (which is in turn related to #24068).

**Comment by bdarnell** [comment]
Yes, exactly. We're concerned about election storms, but for the number of messages rather than for the disruption of existing leaders.

**Comment by tschottdorf** [comment]
Ack. Let's see if something like https://github.com/cockroachdb/cockroach/pull/28581 lands before we do something about dormancy.

**Comment by tbg** [comment]
I checked that v2.0.0 didn't have any of the known bugs (ignoring PreVote candidacy, etc), so I don't know what was going on here.

I ran the roachtest I wrote in #32845 in a mode where it leaves n3 decommissioned (just skip the recommissioning step) and the replica count drops to zero quickly, so I have some confidence that this isn't some obvious problem (though perhaps we did fix something from v2.0.0 to master).
